catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/RoleRepository.java

```diff
@@ -340,8 +338,97 @@ public RoleUpdater(Role original, Role updated, Operation operation) {
+      updateDefault(original.getEntity(), updated.getEntity());
+    }
+
+    private void updateDefault(Role origRole, Role updatedRole) throws IOException, ParseException {
+      long startTime = System.nanoTime();
+      if (Boolean.FALSE.equals(origRole.getDefault()) && Boolean.TRUE.equals(updatedRole.getDefault())) {
+        setDefaultToTrue(updatedRole);
+      }
+      if (Boolean.TRUE.equals(origRole.getDefault()) && Boolean.FALSE.equals(updatedRole.getDefault())) {
+        setDefaultToFalse(updatedRole);
+      }
+      recordChange("default", origRole.getDefault(), updatedRole.getDefault());
+      LOG.info(
+          "Took {} ns to update {} role field default from {} to {}",
+          System.nanoTime() - startTime,
+          updatedRole.getName(),
+          origRole.getDefault(),
+          updatedRole.getDefault());
+    }
+
+    private void setDefaultToTrue(Role role) throws IOException, ParseException {
+      List<Role> defaultRoles = getDefaultRoles(null, ROLE_PATCH_FIELDS);
+      EntityRepository<Role> roleRepository = Entity.getEntityRepository(Entity.ROLE);
+      // Set default=FALSE for all existing default roles.
+      for (Role defaultRole : defaultRoles) {
+        if (defaultRole.getId().equals(role.getId())) {
+          // Skip the current role which is being set with default=TRUE.
+          continue;
+        }
+        Role origDefaultRole = roleRepository.get(null, defaultRole.getId().toString(), ROLE_PATCH_FIELDS);
+        Role updatedDefaultRole = roleRepository.get(null, defaultRole.getId().toString(), ROLE_PATCH_FIELDS);
+        updatedDefaultRole = updatedDefaultRole.withDefault(false);
+        new RoleUpdater(origDefaultRole, updatedDefaultRole, Operation.PATCH).update();
+      }
+      LOG.info("Creating 'user --- has ---> role' relationship for {} role", role.getName());
+      updateUsers(new RoleEntityInterface(role).getEntityReference(), null);
+    }
+
+    private void setDefaultToFalse(Role role) {
+      LOG.info("Deleting 'user --- has ---> role' relationship for {} role", role.getName());
+      updateUsers(null, new RoleEntityInterface(role).getEntityReference());
+    }
+
+    private List<User> getAllUsers() {
+      EntityRepository<User> userRepository = Entity.getEntityRepository(Entity.USER);
+      try {
+        return userRepository
+            .listAfter(null, UserRepository.USER_UPDATE_FIELDS, null, Integer.MAX_VALUE - 1, null, Include.ALL)
+            .getData();
+      } catch (GeneralSecurityException | IOException | ParseException e) {
+        throw EntityNotFoundException.byMessage(CatalogExceptionMessage.entitiesNotFound(Entity.USER));
+      }
+    }
+
+    private void updateUsers(EntityReference addRole, EntityReference removeRole) {
+      List<User> users = getAllUsers();
+      if (users.size() == 0) {
+        return;
+      }
+      EntityRepository<User> userRepository = Entity.getEntityRepository(Entity.USER);
+      users
+          .parallelStream()
```

Comment:
Parallelization to minimize performance impact.

Switching a default role for 100 users takes ~2 seconds on Mattbook Pro


Considered bulk update of relationships in entity_relationship table. 
However, this does not capture the change description for the role update for each of the users. 

Optimizing for correctness instead of performance for this feature.